File: Part2_강의자료/Ch01_큐/문제별코드/10866_덱/10866.py
import sys
input = sys.stdin.readline
from collections import deque

N = int(input())
q = deque()

for i in range(N):
    cmd = input().strip().split()
    if cmd[0] == "push_front":
        q.appendleft(int(cmd[1]))
    elif cmd[0] == "push_back":
        q.append(int(cmd[1]))
    elif cmd[0] == "pop_front":
        print(q.popleft() if q else -1)
    elif cmd[0] == "pop_back":
        print(q.pop() if q else -1)
    elif cmd[0] == "size":
        print(len(q))
    elif cmd[0] == "empty":
        print(1 if not q else 0)
    elif cmd[0] == "front":
        print(q[0] if q else -1)
    elif cmd[0] == "back":
        print(q[-1] if q else -1)

File: Part2_강의자료/Ch01_큐/문제별코드/10866_덱/10866.cpp
#include <stdio.h>
#include <string.h>
#include <deque>
using namespace std;

int main()
{
    int N;
    deque<int> q;
    char cmd[11];
    int value;

    scanf("%d", &N);
    while (N--)
    {
        scanf("%s", cmd);
        if (strcmp(cmd, "push_front") == 0)
        {
            scanf("%d", &value);
            q.push_front(value);
        }
        else if (strcmp(cmd, "push_back") == 0)
        {
            scanf("%d", &value);
            q.push_back(value);
        }
        else if (strcmp(cmd, "pop_front") == 0)
        {
            printf("%d\n", q.empty() ? -1 : q.front());
            if (!q.empty()) q.pop_front();
        }
        else if (strcmp(cmd, "pop_back") == 0)
        {
            printf("%d\n", q.empty() ? -1 : q.back());
            if (!q.empty()) q.pop_back();
        }
        else if (strcmp(cmd, "size") == 0)
            printf("%d\n", q.size());
        else if (strcmp(cmd, "empty") == 0)
            printf("%d\n", q.empty() ? 1 : 0);
        else if (strcmp(cmd, "front") == 0)
            printf("%d\n", q.empty() ? -1 : q.front());
        else if (strcmp(cmd, "back") == 0)
            printf("%d\n", q.empty() ? -1 : q.back());
    }
    return 0;
}
